File: src/main/scala/org/jacop/examples/scala/Nonogram.scala
package org.jacop.examples.scala

import java.io.BufferedReader
import java.io.FileNotFoundException
import java.io.FileReader
import java.io.IOException
import java.util.regex.Pattern

import scala.collection.mutable.ArrayBuffer
import org.jacop.scala._

/**
 * 
 * It solves a nonogram example problem, sometimes also called Paint by Numbers.
 * 
 * 
 * @author Radoslaw Szymanek; rewriting to Scala Krzysztof Kuchcinski
 * 
 */

object Nonogram extends jacop {

  /**
   * The value that represents a black dot.
   */
  val black = 1
	
  /**
   * The value that represents a white dot.
   */
  val white = 0

  /**
   * A board to be painted in white/black dots.
   */
  var board : Array[Array[IntVar]] = null

  /**
   * It specifies if the slide based decomposition of the regular constraint
   * should be applied. This decomposition uses ternary extensional support 
   * constraints. It achieves GAC if FSM is deterministic. 
   */
  val slideDecomposition = false

  /**
   * It specifies if the regular constraint should be used.
   */
  val regularConstr = true

  /**
   * It specifies if one extensional constraint based on MDD created from FSM
   * should be used. The translation process works if FSM is deterministic.
   */
  val extensionalMDD = false

  def readFromFile(filename: String) {
		
    var lines = new Array[String](100)

    val dimensions = new Array[Int](2)

    /* read from file args[0] or qcp.txt */
    try {
      
      val in = new BufferedReader(new FileReader(filename))

      var str = in.readLine()

      val pat = Pattern.compile(" ")
      val result = pat.split(str)
			
      var current = 0

      for (j <- 0 until result.length)
	try {
	  val currentNo = result(j).toInt
	  dimensions(current) = currentNo
	  current += 1
	} catch {
	  case ex : Exception =>
	}

      lines = new Array[String](dimensions(0) + dimensions(1))
			
      var n = 0

      str = in.readLine()	
      while (str != null && n < lines.length) {
	lines(n) = str
	n += 1
	str = in.readLine()
      }
      in.close()
    } catch {
      case e : FileNotFoundException => System.err.println("I can not find file " + filename)
      case e : IOException => {
	System.err.println("Something is wrong with file" + filename)
      }
 
      row_rules = new Array[Array[Int]](dimensions(1)) 
      col_rules = new Array[Array[Int]](dimensions(0))

      // Transforms strings into ints
      for (i <- 0 until lines.length) {
			
	val pat = Pattern.compile(" ")
	val result = pat.split(lines(i))

	val sequence = new Array[Int](result.length)
			
	var current = 0
	for (j <- 0 until result.length)
	  try {
	    sequence(current) = result(j).toInt
	    current += 1
	  } catch {
	    case ex : Exception =>
	  }
	if (i < row_rules.length) row_rules(i) = sequence
	else
	  col_rules(i - row_rules.length) = sequence
      }
      
    }
  }
	

  /**
   * It produces and FSM given a sequence representing a rule. e.g. [2, 3]
   * specifies that there are two black dots followed by three black dots.
   * 
   * @param sequence - input parameter
   * @return Finite State Machine used by Regular automaton to enforce proper sequence.
   */
  def createAutomaton(sequence: Array[Int]) : fsm = {

    var result = new fsm

    var currentState = new state
    result.init(currentState)

    currentState -> (white, currentState)
		
    for (i <- 0 until sequence.length) {
      if (sequence(i) != 0) {
	for (j <- 0 until sequence(i)) {
	  // Black transition
 	  val nextState = new state
 	  result += nextState
 	  currentState -> (black, nextState)
 	  currentState = nextState
	}
	// White transitions
	if (i + 1 != sequence.length) {
	  val nextState = new state
	  result += nextState
	  currentState -> (white, nextState)
	  currentState = nextState
	}  
	currentState -> (white, currentState)
      }
    }

    result.addFinalStates( Array(currentState) )

     result

  }

  def model() {

    import org.jacop.constraints.ExtensionalSupportMDD
    import org.jacop.constraints.regular.Regular

    var vars = new ArrayBuffer[IntVar]()

    var values : IntSet = new IntSet
    values = values + white + black
 
    // Specifying the board with allowed values.
    board = Array.tabulate(row_rules.length, col_rules.length)( (i, j) => 
                       new IntVar("board[" + i + "][" + j + "]", values))

    // Zigzag based variable ordering. 
    for (m <- 0 until row_rules.length + col_rules.length - 1) {
      for (j <- 0 until m if j < col_rules.length) {
	val i = m - j
   	if (i < row_rules.length)
   	  vars :+= board(i)(j)
      }
    }		
		
    println("Size " + vars.length)
		
    // Making sure that rows respect the rules.
    for (i <- 0 until row_rules.length) {
			
      val result = createAutomaton(row_rules(i))
			
      if (slideDecomposition)
  	getModel.imposeDecomposition(new Regular(result, board(i).asInstanceOf[Array[org.jacop.core.IntVar]]))
      
      if (regularConstr)
	regular(result, board(i).toList)
			
      if (extensionalMDD)
 	getModel.impose(new ExtensionalSupportMDD(result.transformDirectlyIntoMDD(board(i).asInstanceOf[Array[org.jacop.core.IntVar]])))

    }

    // Making sure that columns respect the rules.
    for (i <- 0 until col_rules.length) {
					
      val result = createAutomaton(col_rules(i))
      val column = Array.tabulate(row_rules.length)( j => board(j)(i))
							
      if (slideDecomposition)
 	getModel.imposeDecomposition(new Regular(result, column.asInstanceOf[Array[org.jacop.core.IntVar]]))

      if (regularConstr)
	regular(result, column.toList)

      if (extensionalMDD)
 	getModel.impose(new ExtensionalSupportMDD(result.transformDirectlyIntoMDD(column.asInstanceOf[Array[org.jacop.core.IntVar]])))
	
    }

    satisfy( search(vars.toList, input_order, indomain_min) )

  }
			
  /**
   * It prints a matrix of variables. All variables must be grounded.
   * @param matrix matrix containing the grounded variables.
   */
  def printMatrix(matrix: Array[Array[IntVar]]) {

        for(i <- 0 until matrix.length) {
            for(j <- 0 until matrix(i).length) {
            	if ( matrix(i)(j).value() == black )
            		print("0")
            	else
            		print(" ")
            }
            println()
        }
    }

  /**
   * It executes the program which solves this simple problem.
   * @param args no arguments are read.
   */
  def main(args : Array[String]) {

    model()
    printMatrix(board)

/*
    for (i <- 0 until 150) {
			
      var no = ""+i
      while (no.length() < 3)
      no = "0" + no;
      
      System.out.println("Problem file data" + no + ".nin");
      readFromFile("/Users/kris/research/JaCoP-3.1/ExamplesJaCoP/nonogramRepository/data" + no + ".nin");
      model();				
    }
*/
  } 

  /**
   * It specifies a rule for each row.
   */

  var row_rules = Array( 
		  Array(0,0,0,0,2,2,3),
		  Array(0,0,4,1,1,1,4),
		  Array(0,0,4,1,2,1,1),
		  Array(4,1,1,1,1,1,1),
		  Array(0,2,1,1,2,3,5),
		  Array(0,1,1,1,1,2,1),
		  Array(0,0,3,1,5,1,2),
		  Array(0,3,2,2,1,2,2),
		  Array(2,1,4,1,1,1,1),
		  Array(0,2,2,1,2,1,2),
		  Array(0,1,1,1,3,2,3),
		  Array(0,0,1,1,2,7,3),
		  Array(0,0,1,2,2,1,5),
		  Array(0,0,3,2,2,1,2),
		  Array(0,0,0,3,2,1,2),
		  Array(0,0,0,0,5,1,2),
		  Array(0,0,0,2,2,1,2),
		  Array(0,0,0,4,2,1,2),
		  Array(0,0,0,6,2,3,2),
		  Array(0,0,0,7,4,3,2),
		  Array(0,0,0,0,7,4,4),
		  Array(0,0,0,0,7,1,4),
		  Array(0,0,0,0,6,1,4),
		  Array(0,0,0,0,4,2,2),
		  Array(0,0,0,0,0,2,1)
	)
	
  /**
   * It specifies a rule for each column.
   */
	 
  var col_rules = Array( 
		   Array(0,0,1,1,2,2),
		   Array(0,0,0,5,5,7),
		   Array(0,0,5,2,2,9),
		   Array(0,0,3,2,3,9),
		   Array(0,1,1,3,2,7),
		   Array(0,0,0,3,1,5),
		   Array(0,7,1,1,1,3),
		   Array(1,2,1,1,2,1),
		   Array(0,0,0,4,2,4),
		   Array(0,0,1,2,2,2),
		   Array(0,0,0,4,6,2),
		   Array(0,0,1,2,2,1),
		   Array(0,0,3,3,2,1),
		   Array(0,0,0,4,1,15),
		   Array(1,1,1,3,1,1),
		   Array(2,1,1,2,2,3),
		   Array(0,0,1,4,4,1),
		   Array(0,0,1,4,3,2),
		   Array(0,0,1,1,2,2),
		   Array(0,7,2,3,1,1),
		   Array(0,2,1,1,1,5),
		   Array(0,0,0,1,2,5),
		   Array(0,0,1,1,1,3),
		   Array(0,0,0,4,2,1),
		   Array(0,0,0,0,0,3)
	)


	
/*  
  var row_rules = Array(
	Array(2),
	Array(2),
	Array(2),
	Array(2),
	Array(12),
	Array(2),
	Array(2),
	Array(2),
	Array(2),
	Array(2),
	Array(2),
	Array(2),
	Array(2),
	Array(2),
	Array(2),
	Array(2),
	Array(14),
	Array(14),
	Array(1,2),
	Array(1,2),
	Array(1,2),
	Array(1,2),
	Array(11,2),
	Array(1,2),
	Array(1,2),
	Array(1,2),
	Array(1,2),
	Array(1,5),
	Array(1,5),
	Array(1,5),
	Array(1,5),
	Array(1,5),
	Array(1,5),
	Array(1,24),
	Array(32),
	Array(26),
	Array(26),
	Array(26),
	Array(3,8,7,2),
	Array(3,7,6,2),
	Array(3,7,6,2),
	Array(3,7,6,2),
	Array(3,8,7,2),
	Array(26),
	Array(34),
	Array(34),
	Array(34),
	Array(34),
	Array(34),
	Array(24),
	Array(5,3,5),
	Array(3,5,3,5),
	Array(2,2,24),
	Array(1,1,24),
	Array(2,29),
	Array(5,24,2),
	Array(3,26,2),
	Array(41,3),
	Array(43,2),
	Array(42,3),
	Array(41,2),
	Array(41,2),
	Array(4,2,2,2,2,5),
	Array(42),
	Array(42),
	Array(42),
	Array(64),
	Array(86),
	Array(83),
	Array(65),
	Array(75),
	Array(71),
	Array(60,10),
	Array(6,53,10),
	Array(6,25,24,10),
	Array(19,17,3,44),
	Array(30,12,40),
	Array(12,19,22),
	Array(9,24,16),
	Array(4,20,8,11),
	Array(5,24,6,2,3),
	Array(38,6,2,6),
	Array(45,8,8),
	Array(52,3),
	Array(51,1),
	Array(51,2),
	Array(50,1),
	Array(50,1),
	Array(50,1),
	Array(34,9,2),
	Array(31,6,7,7,1),
	Array(29,10,6,3,5,1),
	Array(27,8,5,5,3,1),
	Array(25,5,5,4,10,5,2),
	Array(23,3,11,2,3,3,5,1),
	Array(20,3,16,2,8,11,1),
	Array(18,2,8,4,1,9,3,7,2),
	Array(15,4,8,5,4,12,5,9,1),
	Array(11,5,7,8,2,14,5,11,1),
	Array(7,7,6,13,9,15,15));
	
  var col_rules = Array(
	Array(2),
	Array(4),
	Array(8),
	Array(2,13),
	Array(2,17),
	Array(2,18),
	Array(2,19),
	Array(2,19),
	Array(2,19),
	Array(1,2,19),
	Array(2,2,19),
	Array(2,2,19),
	Array(2,2,19),
	Array(2,2,19),
	Array(2,2,19),
	Array(2,2,18),
	Array(2,26,1),
	Array(2,9,17,1),
	Array(2,10,17,1),
	Array(2,10,16,2),
	Array(2,9,17,2),
	Array(2,9,17,2),
	Array(2,3,3,16,3),
	Array(13,16,2),
	Array(13,15,1),
	Array(12,16,2,1),
	Array(12,15,2,1),
	Array(12,15,1,2),
	Array(11,15,2,3),
	Array(11,15,1,3),
	Array(11,14,2,4),
	Array(1,3,20,14,1,3),
	Array(1,1,2,23,13,1,4),
	Array(1,1,1,21,14,2,4),
	Array(1,1,5,2,22,13,3,3,1),
	Array(1,1,5,4,5,13,13,2,3,1),
	Array(1,1,5,2,5,13,12,2,3,1),
	Array(31,1,6,13,12,3,3,2),
	Array(1,15,1,6,13,12,3,2,2),
	Array(1,29,13,12,3,2,3),
	Array(1,5,32,13,3,2,3),
	Array(1,42,13,2,3,3),
	Array(1,29,12,13,2,3,3),
	Array(29,12,14,1,3,3),
	Array(2,17,10,12,4,8,3,2),
	Array(1,2,17,10,12,3,8,4,1),
	Array(1,2,17,10,12,3,2,9,4,1),
	Array(1,2,17,23,3,2,10,3),
	Array(1,2,6,8,22,3,2,11,3),
	Array(1,2,11,19,11,4,2,11,3),
	Array(38,19,11,4,2,13,1),
	Array(38,19,11,25),
	Array(1,2,11,7,10,12,1,1,3),
	Array(1,2,12,8,10,12,1,1,5),
	Array(1,2,17,23,1,1,7),
	Array(1,2,17,24,1,1,7),
	Array(1,2,17,10,13,1,1,8),
	Array(2,17,10,13,1,1,3,5),
	Array(29,13,4,3,4),
	Array(29,14,2,4,4),
	Array(5,19,14,1,2,4),
	Array(29,14,1,1,3),
	Array(44,2,2,2),
	Array(5,20,2,3,1),
	Array(5,5,14,2,2,2),
	Array(5,5,15,3,2,1),
	Array(5,5,15,3,3,1),
	Array(5,5,15,6,1),
	Array(5,15,3,3),
	Array(21,3,3),
	Array(21,3,3),
	Array(22,2,2),
	Array(2,17,2,2),
	Array(2,2,13,3,1),
	Array(1,2,13,2,1),
	Array(2,2,13,2,1),
	Array(2,1,6,5,4),
	Array(1,2,14,4),
	Array(1,14,4),
	Array(14,2),
	Array(14,2),
	Array(14,3),
	Array(14,2),
	Array(14,2),
	Array(15,2),
	Array(15,2),
	Array(15,2),
	Array(5,2,1,1),
	Array(2,1,2,1,1),
	Array(2,1,2,1,1),
	Array(2,1,2,1,1),
	Array(2,1,2,2,1),
	Array(2,1,2,2,1),
	Array(1,2,1,1),
	Array(1,2,1,1),
	Array(2,1,4),
	Array(2,1,4),
	Array(2,2,5),
	Array(2,2,5),
	Array(3));
*/
/*
  var row_rules = Array(
			  Array(3),
			  Array(5),
			  Array(3,1),
			  Array(2,1),
			  Array(3,3,4),
			  Array(2,2,7),
			  Array(6,1,1),
			  Array(4,2,2),
			  Array(1,1),
			  Array(3,1),
			  Array(6),
			  Array(2,7),
			  Array(6,3,1),
			  Array(1,2,2,1,1),
			  Array(4,1,1,3),
			  Array(4,2,2),
			  Array(3,3,1),
			  Array(3,3),
			  Array(3),
			  Array(2,1)
			);

  var col_rules = 
	 Array(
	  Array(2),
	  Array(1,2),
	  Array(2,3),
	  Array(2,3),
	  Array(3,1,1),
	  Array(2,1,1),
	  Array(1,1,1,2,2),
	  Array(1,1,3,1,3),
	  Array(2,6,4),
	  Array(3,3,9,1),
	  Array(5,3,2),
	  Array(3,1,2,2),
	  Array(2,1,7),
	  Array(3,3,2),
	  Array(2,4),
	  Array(2,1,2),
	  Array(2,2,1),
	  Array(2,2),
	  Array(1),
	  Array(1)
	 );
*/

	/*
	public int[][] col_rules = Array( Array(3), Array(1 ,1), Array(1,1), Array(2));
	public int[][] row_rules = Array( Array(4), Array(1,1), Array(2), Array(1));
    */	
	
}

File: jacop-examples-java/src/main/java/org/jacop/examples/fd/nonogram/Nonogram.java
/*
 * Nonogram.java
 * This file is part of JaCoP.
 * <p>
 * JaCoP is a Java Constraint Programming solver.
 * <p>
 * Copyright (C) 2000-2026 Krzysztof Kuchcinski and Radoslaw Szymanek
 * <p>
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * <p>
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 * <p>
 * Notwithstanding any other provision of this License, the copyright
 * owners of this work supplement the terms of this License with terms
 * prohibiting misrepresentation of the origin of this work and requiring
 * that modified versions of this work be marked in reasonable ways as
 * different from the original version. This supplement of the license
 * terms is in accordance with Section 7 of GNU Affero General Public
 * License version 3.
 * <p>
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see http://www.gnu.org/licenses/.
 */

package org.jacop.examples.fd.nonogram;

import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.regex.Pattern;
import org.jacop.constraints.ExtensionalSupportMdd;
import org.jacop.constraints.regular.Regular;
import org.jacop.core.IntDomain;
import org.jacop.core.IntVar;
import org.jacop.core.IntervalDomain;
import org.jacop.core.Store;
import org.jacop.examples.fd.ExampleFd;
import org.jacop.search.DepthFirstSearch;
import org.jacop.search.IndomainMin;
import org.jacop.search.InputOrderSelect;
import org.jacop.search.SelectChoicePoint;
import org.jacop.util.fsm.Fsm;
import org.jacop.util.fsm.FsmState;
import org.jacop.util.fsm.FsmTransition;

/**
 * It solves a nonogram example problem, sometimes also called Paint by Numbers.
 *
 * @author Radoslaw Szymanek
 * @version 5.0
 */
public class Nonogram extends ExampleFd {

  /** The value that represents a black dot. */
  public final int black = 1;

  /** The value that represents a white dot. */
  public final int white = 0;

  /**
   * It specifies if the slide based decomposition of the regular constraint should be applied. This
   * decomposition uses ternary extensional support constraints. It achieves GAC if Fsm is
   * deterministic.
   */
  public final boolean slideDecomposition = false;

  /** It specifies if the regular constraint should be used. */
  public final boolean regular = true;

  /**
   * It specifies if one extensional constraint based on Mdd created from Fsm should be used. The
   * translation process works if Fsm is deterministic.
   */
  public final boolean extensionalMdd = false;

  /** A board to be painted in white/black dots. */
  public IntVar[][] board;

  private static final String SOLUTION_FOUND = "Solution(s) found";

  /** It specifies a rule for each row. */
  public int[][] row_rules = {
    {0, 0, 0, 0, 2, 2, 3},
    {0, 0, 4, 1, 1, 1, 4},
    {0, 0, 4, 1, 2, 1, 1},
    {4, 1, 1, 1, 1, 1, 1},
    {0, 2, 1, 1, 2, 3, 5},
    {0, 1, 1, 1, 1, 2, 1},
    {0, 0, 3, 1, 5, 1, 2},
    {0, 3, 2, 2, 1, 2, 2},
    {2, 1, 4, 1, 1, 1, 1},
    {0, 2, 2, 1, 2, 1, 2},
    {0, 1, 1, 1, 3, 2, 3},
    {0, 0, 1, 1, 2, 7, 3},
    {0, 0, 1, 2, 2, 1, 5},
    {0, 0, 3, 2, 2, 1, 2},
    {0, 0, 0, 3, 2, 1, 2},
    {0, 0, 0, 0, 5, 1, 2},
    {0, 0, 0, 2, 2, 1, 2},
    {0, 0, 0, 4, 2, 1, 2},
    {0, 0, 0, 6, 2, 3, 2},
    {0, 0, 0, 7, 4, 3, 2},
    {0, 0, 0, 0, 7, 4, 4},
    {0, 0, 0, 0, 7, 1, 4},
    {0, 0, 0, 0, 6, 1, 4},
    {0, 0, 0, 0, 4, 2, 2},
    {0, 0, 0, 0, 0, 2, 1}
  };

  /** It specifies a rule for each column. */
  public int[][] col_rules = {
    {0, 0, 1, 1, 2, 2},
    {0, 0, 0, 5, 5, 7},
    {0, 0, 5, 2, 2, 9},
    {0, 0, 3, 2, 3, 9},
    {0, 1, 1, 3, 2, 7},
    {0, 0, 0, 3, 1, 5},
    {0, 7, 1, 1, 1, 3},
    {1, 2, 1, 1, 2, 1},
    {0, 0, 0, 4, 2, 4},
    {0, 0, 1, 2, 2, 2},
    {0, 0, 0, 4, 6, 2},
    {0, 0, 1, 2, 2, 1},
    {0, 0, 3, 3, 2, 1},
    {0, 0, 0, 4, 1, 15},
    {1, 1, 1, 3, 1, 1},
    {2, 1, 1, 2, 2, 3},
    {0, 0, 1, 4, 4, 1},
    {0, 0, 1, 4, 3, 2},
    {0, 0, 1, 1, 2, 2},
    {0, 7, 2, 3, 1, 1},
    {0, 2, 1, 1, 1, 5},
    {0, 0, 0, 1, 2, 5},
    {0, 0, 1, 1, 1, 3},
    {0, 0, 0, 4, 2, 1},
    {0, 0, 0, 0, 0, 3}
  };

  /**
   * It executes the program which solves this simple problem.
   *
   * @param args no arguments are read.
   */
  static void main(String[] args) {

    Nonogram example = new Nonogram();

    example.model();
    if (example.searchAll()) {
      IO.println(SOLUTION_FOUND);
    }

    example.printMatrix(example.board);
  }

  /**
   * It executes the program which solves this simple problem.
   *
   * @param args no arguments are read.
   */
  public static void test(String[] args) {

    Nonogram example = new Nonogram();

    example.model();
    if (example.searchAll()) {
      IO.println(SOLUTION_FOUND);
    }
    example.printMatrix(example.board);

    for (int i = 0; i <= 150; i++) {

      StringBuilder no = new StringBuilder(String.valueOf(i));
      while (no.length() < 3) {
        no.insert(0, "0");
      }

      IO.println("Problem file data" + no + ".nin");
      example.readFromFile("ExamplesJaCoP/nonogramRepository/data" + no + ".nin");
      example.model();

      if (example.searchAll()) {
        IO.println(SOLUTION_FOUND);
      }

      example.printMatrix(example.board);
    }
  }

  /** Parses a space-separated line into an int array, ignoring parse errors. */
  private static int[] parseIntSequence(String line) {
    Pattern pat = Pattern.compile(" ");
    String[] result = pat.split(line);
    int[] sequence = new int[result.length];
    int current = 0;
    for (String s : result) {
      try {
        sequence[current++] = Integer.parseInt(s);
      } catch (Exception _) {
        // Ignore parsing errors
      }
    }
    return sequence;
  }

  /**
   * Reads a nonogram problem definition from a file.
   *
   * @param filename path to the file containing the nonogram puzzle specification
   */
  public void readFromFile(String filename) {

    String[] lines = new String[100];
    int[] dimensions = new int[2];

    try (BufferedReader in =
        new BufferedReader(
            new InputStreamReader(new FileInputStream(filename), StandardCharsets.UTF_8))) {

      String str = in.readLine();
      Pattern pat = Pattern.compile(" ");
      String[] result = pat.split(str != null ? str : "");
      int current = 0;
      for (String s : result) {
        try {
          dimensions[current++] = Integer.parseInt(s);
        } catch (Exception _) {
          // Ignore parsing errors
        }
      }

      lines = new String[dimensions[0] + dimensions[1]];
      int n = 0;
      while ((str = in.readLine()) != null && n < lines.length) {
        lines[n] = str;
        n++;
      }
    } catch (FileNotFoundException _) {
      System.err.println("I can not find file " + filename);
    } catch (IOException _) {
      System.err.println("Something is wrong with file" + filename);
    }

    row_rules = new int[dimensions[1]][];
    col_rules = new int[dimensions[0]][];

    for (int i = 0; i < lines.length; i++) {
      int[] sequence = parseIntSequence(lines[i]);
      if (i < row_rules.length) {
        row_rules[i] = sequence;
      } else {
        col_rules[i - row_rules.length] = sequence;
      }
    }
  }

  /**
   * It produces and Fsm given a sequence representing a rule. e.g. [2, 3] specifies that there are
   * two black dots followed by three black dots.
   *
   * @param sequence a sequence representing a rule. e.g. [2, 3]
   * @return Finite State Machine used by Regular automaton to enforce proper sequence.
   */
  public Fsm createAutomaton(int[] sequence) {

    Fsm result = new Fsm();

    FsmState currentState = new FsmState();

    result.initState = currentState;
    IntDomain blackEncountered = new IntervalDomain(black, black);
    IntDomain whiteEncountered = new IntervalDomain(white, white);

    FsmTransition white = new FsmTransition(whiteEncountered, currentState);
    currentState.addTransition(white);

    for (int i = 0; i < sequence.length; i++) {
      if (sequence[i] == 0) {
        continue;
      }
      for (int j = 0; j < sequence[i]; j++) {
        // Black transition
        FsmState nextState = new FsmState();
        FsmTransition black = new FsmTransition(blackEncountered, nextState);
        currentState.addTransition(black);
        result.allStates.add(currentState);
        currentState = nextState;
      }
      // White transitions
      if (i + 1 != sequence.length) {
        FsmState nextState = new FsmState();
        white = new FsmTransition(whiteEncountered, nextState);
        currentState.addTransition(white);
        result.allStates.add(currentState);
        currentState = nextState;
      }

      white = new FsmTransition(whiteEncountered, currentState);
      currentState.addTransition(white);
    }

    result.allStates.add(currentState);
    result.finalStates.add(currentState);

    return result;
  }

  @Override
  public void model() {

    store = new Store();
    vars = new ArrayList<>();
    initBoardAndValues();
    addZigzagVariableOrdering();
    IO.println("Size " + vars.size());
    addRowRules();
    addColumnRules();
  }

  private void initBoardAndValues() {
    IntervalDomain values = new IntervalDomain();
    values.unionAdapt(black, black);
    values.unionAdapt(white, white);
    board = new IntVar[row_rules.length][col_rules.length];
    for (int i = 0; i < board.length; i++) {
      for (int j = 0; j < board[0].length; j++) {
        board[i][j] = new IntVar(store, "board[" + i + "][" + j + "]", values.copy());
      }
    }
  }

  private void addZigzagVariableOrdering() {
    for (int m = 0; m < row_rules.length + col_rules.length - 1; m++) {
      for (int j = 0; j <= m && j < col_rules.length; j++) {
        int i = m - j;
        if (i >= row_rules.length) {
          continue;
        }
        vars.add(board[i][j]);
      }
    }
  }

  private void addRowRules() {
    for (int i = 0; i < row_rules.length; i++) {
      Fsm result = this.createAutomaton(row_rules[i]);
      imposeRowConstraint(result, board[i]);
    }
  }

  private void imposeRowConstraint(Fsm result, IntVar[] row) {
    if (slideDecomposition) {
      store.imposeDecomposition(new Regular(result, row));
    }
    if (regular) {
      store.impose(new Regular(result, row));
    }
    if (extensionalMdd) {
      store.impose(new ExtensionalSupportMdd(result.transformDirectlyIntoMdd(row)));
    }
  }

  private void addColumnRules() {
    for (int i = 0; i < col_rules.length; i++) {
      Fsm result = createAutomaton(col_rules[i]);
      IntVar[] column = new IntVar[row_rules.length];
      for (int j = 0; j < column.length; j++) {
        column[j] = board[j][i];
      }
      imposeRowConstraint(result, column);
    }
  }

  /**
   * It specifies simple search method based on most constrained static and lexigraphical ordering
   * of values. It searches for all solutions.
   *
   * @return true if there is a solution, false otherwise.
   */
  public boolean searchAll() {

    // In case of nonograms, value ordering does not matter since we
    // a) search for all solutions
    // b) all variables have binary domain.
    final SelectChoicePoint<IntVar> select =
        new InputOrderSelect<>(store, vars.toArray(new IntVar[1]), new IndomainMin<>());

    searchLabel = new DepthFirstSearch<>();

    searchLabel.getSolutionListener().searchAll(true);
    searchLabel.getSolutionListener().recordSolutions(false);
    searchLabel.setAssignSolution(true);

    IO.println("Search has begun ...");

    final long T1 = System.currentTimeMillis();

    boolean result = searchLabel.labeling(store, select);

    long T2 = System.currentTimeMillis();

    if (result) {
      IO.println("Number of solutions " + searchLabel.getSolutionListener().solutionsNo());
      searchLabel.printAllSolutions();
    } else {
      IO.println("Failed to find any solution");
    }

    IO.println("\n\t*** Execution time = " + (T2 - T1) + " ms");

    return result;
  }

  /**
   * It prints a matrix of variables. All variables must be grounded.
   *
   * @param matrix matrix containing the grounded variables.
   */
  public void printMatrix(IntVar[][] matrix) {

    for (IntVar[] intVars : matrix) {
      for (IntVar intVar : intVars) {
        if (intVar.value() == black) {
          IO.print("0");
        } else {
          IO.print(" ");
        }
      }
      IO.println();
    }
  }
}
